### PR TITLE
Add hardware stats to train_head

### PR DESCRIPTION
**Pull request by alanwguo**
## Why are these changes needed?
CPU utilization and GPU utilization is useful to see next to Train Workers.
Fixes actor GPU utilization not working due to bug introduced in #41399

Also refactor to use DataOrganizer to re-use more code with rest of dashboard.

Screenshot 2024-07-23 at 1 24 17 AM

## Related issue number
fixes bug introduced in #41399

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by scottsun94** [comment]
nice!

File: python/ray/dashboard/modules/train/train_head.py [hunk 1]
```diff
@@ -95,43 +91,85 @@ async def get_train_runs(self, req: Request) -> Response:
...
-                        "state"
-                    ]
+            actors = await DataOrganizer.get_all_actors()
+
         except Exception:
             logger.exception("Error Getting all actor info from GCS.")
 
+        train_runs_with_details: List[TrainRunInfoWithDetails] = []
+
         for train_run in train_runs.values():
```
Comment (woshiyyya) [inline python/ray/dashboard/modules/train/train_head.py:110]: Probably we can skip updating the hardware information for the finished train_runs to make it more efficient.

File: python/ray/train/_internal/state/schema.py [hunk 2]
```diff
@@ -91,6 +140,9 @@ class TrainRunInfo(BaseModel):
 class TrainRunInfoWithDetails(TrainRunInfo):
     """Metadata for a Ray Train run and information about its workers."""
 
+    workers: List[TrainWorkerInfoWithDetails] = Field(
```
Comment (woshiyyya) [inline python/ray/train/_internal/state/schema.py:143]: Will this override the `workers` field in the parent class (`TrainRunInfo`)?

**Review by woshiyyya — COMMENTED** [review]
Thank you!! Left some comments

File: python/ray/train/_internal/state/schema.py [hunk 3]
```diff
@@ -47,6 +47,55 @@ class TrainWorkerInfo(BaseModel):
     )
 
 
+@DeveloperAPI
+class MemoryInfo(BaseModel):
```
[the remark this answers is not in the document]
Comment (alanwguo) [reply python/ray/train/_internal/state/schema.py:51]: I looked around and I don't think so.

File: python/ray/dashboard/modules/train/train_head.py [hunk 4]
```diff
@@ -95,43 +91,85 @@ async def get_train_runs(self, req: Request) -> Response:
...
+                        for gpu in actor["gpus"]
+                        if worker_info.pid
+                        in [process["pid"] for process in gpu["processesPids"]]
+                    ]
+                    # Need to convert processesPids into a proper list.
+                    # It's some weird ImmutableList structureo
+                    formatted_gpus = [
+                        {
+                            **gpu,
+                            "processesPids": list(gpu["processesPids"]),
```
[the remark this answers is not in the document]
Comment (alanwguo) [reply python/ray/dashboard/modules/train/train_head.py:128]: yes, this is already the case. It's filtered down, I'll try changing this to not be a list.

File: python/ray/dashboard/modules/train/train_head.py (hunk 1 above)
Comment (alanwguo) [reply python/ray/dashboard/modules/train/train_head.py:110]: good idea

File: python/ray/train/_internal/state/schema.py (hunk 2 above)
Comment (alanwguo) [reply python/ray/train/_internal/state/schema.py:143]: yes, that was my intention.

File: python/ray/dashboard/modules/train/train_head.py (hunk 1 above)
Comment (alanwguo) [reply python/ray/dashboard/modules/train/train_head.py:110]: I decided not to do this since the amount of work is neglible

Comment (woshiyyya) [reply python/ray/dashboard/modules/train/train_head.py:110]: oh yes. That makes sense.

**Review by woshiyyya — APPROVED** [review]
The `train_head` and `schema` part looks good to me!

File: python/ray/train/_internal/state/schema.py [hunk 5]
```diff
@@ -47,6 +47,55 @@ class TrainWorkerInfo(BaseModel):
...
+    rss: int
+    vms: int
+    pfaults: Optional[int]
+    pageins: Optional[int]
+
+
+@DeveloperAPI
+class ProcessStats(BaseModel):
```
Comment (matthewdeng) [inline python/ray/train/_internal/state/schema.py:59]: Is this schema based on an existing one? See questions below

File: python/ray/train/_internal/state/schema.py [hunk 6]
```diff
@@ -47,6 +47,55 @@ class TrainWorkerInfo(BaseModel):
...
+    cpuPercent: float
```
Comment (matthewdeng) [inline python/ray/train/_internal/state/schema.py:60]: Will this be 0 to 1 or 0 to 100?

File: python/ray/train/_internal/state/schema.py [hunk 7]
```diff
@@ -47,6 +47,55 @@ class TrainWorkerInfo(BaseModel):
...
+    # total memory, free memory, memory used ratio
+    mem: Optional[List[int]]
+    memoryInfo: MemoryInfo
```
Comment (matthewdeng) [inline python/ray/train/_internal/state/schema.py:63]: Why is `mem` a list and not individual fields? Why are these not part of `memoryInfo`?

File: python/ray/train/_internal/state/schema.py [hunk 8]
```diff
@@ -47,6 +47,55 @@ class TrainWorkerInfo(BaseModel):
...
+
+
+@DeveloperAPI
+class GPUStats(BaseModel):
+    uuid: str
+    index: int
+    name: str
+    utilizationGpu: Optional[float]
+    memoryUsed: float
+    memoryTotal: float
```
Comment (matthewdeng) [inline python/ray/train/_internal/state/schema.py:79]: Should these be in `ProcessGPUUSage`?

File: python/ray/train/_internal/state/schema.py [hunk 9]
```diff
@@ -47,6 +47,55 @@ class TrainWorkerInfo(BaseModel):
...
+    cpuPercent: float
+    # total memory, free memory, memory used ratio
+    mem: Optional[List[int]]
+    memoryInfo: MemoryInfo
+
+
+class ProcessGPUUsage(BaseModel):
+    # This gpu usage stats from a process
+    pid: int
+    gpuMemoryUsage: int
```
Comment (matthewdeng) [inline python/ray/train/_internal/state/schema.py:69]: Would this give the usage specific to the `pid`? In that case should `GPUStats` actually take a list of `processInfo`s?

File: python/ray/train/_internal/state/schema.py (hunk 5 above)
Comment (alanwguo) [reply python/ray/train/_internal/state/schema.py:59]: yeah, this is the existing gpus schema for nodes and actors. I agree it's ugly...

I think with export API, we have a chance to really clean this up for the future but let's not bundle it in with the train dashboard changes.

**Pull request merged**